### DISCUSSION TITLE
TreeControl nodes get underlined and stay that way when hovered over

### DISCUSSION THREAD

**Discussion by jeffwright13** [Textualize/textual#405] [Q&A] [answered]
I have an app that uses the TreeControl widget to display categorized results of a Pytest run. Whenever a node is hovered over, it gets underlined, but the underline stays even after moving the mouse cursor away from the node. Does anyone know if this is a bug in Textual, or (more likely) an issue with [my code](https://github.com/jeffwright13/pytest-tui/blob/main/pytest_tui/tui_textual1.py)?

Here's a quick screencast that shows the issue.

https://user-images.githubusercontent.com/4308435/164020578-12588170-6402-4417-ab48-668dcf741217.mov

**Comment by jeffwright13** [comment]
Not sure if anyone has a clue about this...but it's still an issue for me in 0.1.17.

**Comment by jeffwright13** [comment]
OK so I set up a debugger and set some breakpoints, and I can see where the underline is happening now: 

`/textual/widgets/_tree_control.py`:
```
    if node.id == self.hover_node:
        label.stylize("underline")
```

OK, great, I understand why the 'leaf' of the tree gets underlined. But shouldn't it de-underline itself when the leaf isn't hovered over anymore?

**Comment by rodrigogiraoserrao** [comment]
> **Disclaimer**: I have near-0 knowledge of Textual

The `hover_node` is set when the mouse moves, but *it seems* like nothing is done to the *previous* `hover_node` (when there was one), and that's probably why the style remains. Maybe there could be a check to see if the `hover_node` is changing and clear the styles that no longer apply? Hope this helps.

https://github.com/Textualize/textual/blob/848bd0596689c1801f5a283e5e819f43f0ff6b26/src/textual/widgets/_tree_control.py#L273-L274

**Reply by jeffwright13** [reply]
Thanks for the info. As far as I know Textual doesn't have hooks, so there's no way for me to override this in my own code, I don't think. So because this is builtin behavior, if I wanted to change it, I'd have to submit a PR and wait for a new version of Textual to be cut. Unless someone has some bright ideas? :-)

**Reply by ThatXliner** [reply]
wacky monkey patching?

**Comment by jeffwright13** [comment, marked as answer]
If I edit `_tree_control.py` and I change

```
        if node.id == self.hover_node:
            label.stylize("underline")
```

to

```
        if node.id == self.hover_node:
            label.stylize("underline")
        else:
            label.stylize("not underline")
```
then the underline disappears when the node is not hovered over.

I don't know if this is something @willmcgugan would be interested in incorporating into a new version of Textual. I guess in the meantime I'll see if I can ship my package with an altered version of `_tree_control.py`.

**Reply by rodrigogiraoserrao** [reply]
I would think/guess/hope there is something "smarter" that can be done, in the sense that now you have to stylize every single node every time you render, even those who have never been touched. That may or may not be faster than just "unstylizing" the one that was underlined previously... Just a thought.

**Reply by jeffwright13** [reply]
Agreed - although it might not be worth making a more general solution right now, given that the "official" Textual is going to drop soon, and we don't know what changes might be made there. I mean, there might not even be a TreeControl in the new version, at least not like it exists now.
